### PR TITLE
Panel buffer options

### LINKED ISSUE #1226: Add ability to set detector panel buffer from masks.
We could use a 3rd option for configuring the panel buffer on the detector: set from masks.  I have lost track of the logic of how the various masking options are handled @bnmajor -- but we could make it explicit regarding setting the `panel_buffer` attribute on each detector with the masks defined on that detector.

### PR DESCRIPTION
**Pull request by bnmajor**
Adds some of the features requested in #1226:

- Can clobber any existing panel buffer
- Can do a logical OR to merge with existing
- Can do a logical AND with existing

This will close #1226 (re-opened)

### CODE REVIEW

**Comment by bnmajor** [comment]
Need to:
- [ ] Only provide options dialog if 2-D panel buffer already exists
- [ ] For threshold mask use the current frame

**Review by joelvbernier — APPROVED** [review]

File: hexrd/ui/mask_manager_dialog.py
Comment (psavery) [inline hexrd/ui/mask_manager_dialog.py:343]: FYI: I think you can pretty much do the same thing here with just [`QInputDialog.getItem()`](https://doc.qt.io/qt-5/qinputdialog.html#getItem). But you don't need to change this if it works.

**Comment by psavery** [comment]
Okay, here is my test of this:

For this input image that is thresholded (where blue is thresholded):

thresholded

And this original panel buffer:

original_panel_buffer

When I apply the threshold masks to the panel buffer, here is what I get for the different scenarios:

Replacement
===========
result_of_replacement

Logical Or
========
result_of_or

Logical And
=========
result_of_and

It looks to me like the "And" and "Or" are reversed. Maybe this is because the panel buffer has 1 set to valid pixels and 0 set to invalid pixels. What do you think, @joelvbernier ?

**Comment by joelvbernier** [comment]
Yes @bnmajor and @psavery -- the `panel_buffer` behaves in the oppsite notice of a mask; valid (unmasked) pixels are `True` and invalid (masked) pixels are `False`.  So if the masks are `True` where masked -- like a numpy masked array -- we need to apply the '`~`' operator when setting the detector `panel_buffer` property with a mask.  @bnmajor can you verify the convention of the mask arrays?  We get that sorted and we can merge this up!

**Comment by joelvbernier** [comment]
Actually, this look ok -- OR is adding the extra strip along with the existing buffer, and AND is only adding the pixels that are `True` in both, hence the salt and pepper look.  LGTM!

**Review by psavery — APPROVED** [review]
@joelvbernier confirmed they are correct, so LGTM once all your final edits are pushed @bnmajor

**Comment by bnmajor** [comment]
@joelvbernier This branch has been updated to fix the issues that you were seeing as well as some I noticed. If you and/or @psavery wouldn't mind giving this one more test I think it should be all set now.
- Threshold mask are now applied in Polar View
- Masks were not being toggled on/off in raw view if threshold is visible. This is fixed now (we are no longer updating a copy of the imageseries but rather applying/removing the threshold mask like any other mask).
- Deleting threshold mask when not visible no longer creates error

**Comment by psavery** [comment]
A few things I've noticed that I believe need to be addressed:

- [ ] Toggling the "Set Threshold Mask" checkbox on and off repeatedly does not seem to behave properly. Sometimes, the checkbox is unchecked but the masks are applied, and sometimes, the checkbox is checked but the masks are not applied. The "Visible" checkbox in the mask manager also does not appear to stay in sync with the "Set Threshold Mask" checkbox.
- [ ] The threshold masks appear different in the polar view than in the raw view. This is because the threshold mask is being created after several steps of the polar view processing (image warping, rescaling, and snipping). The pixel values at this point are pretty different than what they were originally, so the thresholding ends up being pretty different as well. I'm wondering if we should perhaps instead create the polar threshold masks by taking the original raw threshold masks and warping them (like is done for the polar images themselves), then applying these. What do you think, @joelvbernier?
- [ ] In the mask manager, if you rename the threshold mask, editing the visibility from the mask manager no longer works properly. Should we disable renaming it, if we are going to treat "threshold" as a key word for the name of the mask? If we do this, we should also not allow other masks to be named "threshold".
- [ ] In the mask manager, if you press the "Remove Mask" button for the threshold mask, it should probably also uncheck "Set Threshold Mask" as well.

**Comment by joelvbernier** [comment]
@psavery -- since the threshold mask is generated in raw image coordinates (pixel coordinates --> cartesian), it should behave the same as the others, which are stored in native cartesian (or pixel) coordinates (right @bnmajor?).  Snipping doesn't affect the masks as it does not alter coordinates, just image intensities.  In theory, masks in raw coordinates get transformed to polar using either the straight `cart_to_angles` on a detector-by-detector basis _or_ using the `warp_image` method on the `PolarView` class.

**Comment by psavery** [comment]
@bnmajor Joel and I talked about converting the threshold masks to polar. I think what we are going to do is to take the raw threshold masks and convert them to the polar image via warping. In #1275, I've added a `PolarView.warp_image()` function that we can use to do this, and I can help set that up after #1275 is merged.

**Comment by psavery** [comment]
LGTM

**Review by joelvbernier — APPROVED** [review]

**Comment by joelvbernier** [comment]
The "equal to" option is not giving the correct results.  in the tardis example, equal to 0 is blanking the whole image.

**Comment by joelvbernier** [comment]
Also need to increase the limits of the QScientificSpinbox to ±2**32

**Comment by joelvbernier** [comment]
Also mask merging with panel buffer does not seem to be working -- I did logical OR with the existing and my panel buffers are now uniformly `True`, which is definitely not the case

**Comment by bnmajor** [comment]
> The "equal to" option is not giving the correct results. in the tardis example, equal to 0 is blanking the whole image.

Fixed in [58a4142](https://github.com/HEXRD/hexrdgui/pull/1281/commits/58a4142826acb1ca5f618201948e4b9b782c1283)

> Also need to increase the limits of the QScientificSpinbox to ±2**32

If this is a request for all `QScientificSpinbox`es this should be a separate PR.

> Also mask merging with panel buffer does not seem to be working -- I did logical OR with the existing and my panel buffers are now uniformly True, which is definitely not the case

This was because the mask that was generated was incorrect, not because of the `logical_or` operation. Fixed in [58a4142](https://github.com/HEXRD/hexrdgui/pull/1281/commits/58a4142826acb1ca5f618201948e4b9b782c1283)

@joelvbernier This branch also includes the `View Masks` option now.

masking

**Pull request merged**
